csrc/hmac.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <openssl/hmac.h>
+#include "env.h"
+#include "util.h"
+#include "buffer.h"
+
+#define DO_NOT_INIT -1
+#define DO_NOT_REKEY -2
+
+using namespace AmazonCorrettoCryptoProvider;
+
+// Some of the logic around how to manage arrays is non-standard because HMAC is extremely performance sensitive.
+// For the smaller data-sizes we're using, avoiding GetPrimitiveArrayCritical is worth it.
+
+namespace {
+    void maybe_init_ctx(raii_env &env, HMAC_CTX *ctx, jbyteArray &keyArr, jlong evpMd) {
+        if (DO_NOT_INIT == evpMd) {
+            return;
+        }
+        if (evpMd == DO_NOT_REKEY)
+        {
+            if (unlikely(HMAC_Init_ex(
+                                ctx,
+                                nullptr, // key
+                                0, // keyLen
+                                nullptr, // EVP_MD
+                                nullptr /* ENGINE */) != 1))
+            {
+                throw_openssl("Unable to initialize HMAC_CTX");
+            }
+        }
+        else
+        {
+            // We pass in keyArr as a jbyteArray to avoid even the minimimal JNI costs
+            // of wrapping it in a java_buffer when we don't need it.
+            java_buffer keyBuf = java_buffer::from_array(env, keyArr);
+            jni_borrow key(env, keyBuf, "key");
+            if (unlikely(HMAC_Init_ex(
+                             ctx,
+                             key.data(),
+                             key.len(),
+                             reinterpret_cast<const EVP_MD *>(evpMd),
+                             nullptr /* ENGINE */) != 1))
+            {
+                throw_openssl("Unable to initialize HMAC_CTX");
+            }
+        }
+    }
+
+    void update_ctx(raii_env &env, HMAC_CTX *ctx, jni_borrow &input) {
+        if (unlikely(HMAC_Update(
+                         ctx,
+                         input.data(),
+                         input.len()) != 1))
+        {
+            throw_openssl("Unable to update HMAC_CTX");
+        }
+    }
+
+    void calculate_mac(raii_env &env, HMAC_CTX *ctx, java_buffer &result) {
+        uint8_t scratch[EVP_MAX_MD_SIZE];
+        unsigned int macSize =EVP_MAX_MD_SIZE;
+        if (unlikely(HMAC_Final(
+                         ctx,
+                         scratch,
+                         &macSize) != 1))
+        {
+            throw_openssl("Unable to update HMAC_CTX");
+        }
+        // When we don't need to read the data in an array but use it strictly for output
+        // it can be faster to use put_bytes rather than convert it into a jni_borrow.
+        result.put_bytes(env, scratch, 0, macSize);
```

Comment:
Getting a pointer to `result`'s underlying array requires entering a critical region in Java. This can be expensive (involving lots of locks, interfacing with the GC, etc.) and for small amounts of data (such as a Mac) it is often faster to just copy the data in rather than deal with critical regions.

Reply:
i see, thanks.